src/evolve.ts

```diff
@@ -0,0 +1,236 @@
+ * Creates return type from the type of arguments of `evolve`.
+ */
+type Evolved<T, E> = E extends AFunction
+  ? ReturnType<E>
+  : T extends object
+    ? {
+        [K in keyof T]: Evolved<T[K], GetValueByKey<E, K, T[K]>>;
+      }
+    : T;
+
+/**
+ * Creates a new object by applying functions that is included in `evolver` object parameter
+ * to the `data` object parameter according to their corresponding path.
+ *
+ * Functions included in `evolver` object will not be invoked
+ * if its corresponding key does not exist in the `data` object.
+ * Also, values included in `data` object will not be used
+ * if its corresponding key does not exist in the `evolver` object.
+ *
+ * @param data object whose value is applied to the corresponding function
+ * that is defined in `evolver` at the same path.
+ * @param evolver object that include functions that is applied to
+ * the corresponding value of `data` object at the same path.
+ * @signature
+ *    R.evolve(data, evolver)
+ * @example
+ *    const transf = {
+ *      count: add(1),
+ *      time: { elapsed: add(1), remaining: add(-1) },
+ *    };
+ *    const data = {
+ *      id: 10,
+ *      count: 10,
+ *      time: { elapsed: 100, remaining: 1400 },
+ *    };
+ *    evolve(data, transf)
+ *    // => {
+ *    //   id: 10,
+ *    //   count: 11,
+ *    //   time: { elapsed: 101, remaining: 1399 },
+ *    // }
+ * @dataFirst
+ * @category Object
+ */
+export function evolve<T, E extends Evolver<T>>(
+  object: T,
+  evolver: E
+): Evolved<T, E>;
+
+/**
+ * Creates a new object by applying functions that is included in `evolver` object parameter
+ * to the `data` object parameter according to their corresponding path.
+ *
+ * Functions included in `evolver` object will not be invoked
+ * if its corresponding key does not exist in the `data` object.
+ * Also, values included in `data` object will not be used
+ * if its corresponding key does not exist in the `evolver` object.
+ *
+ * @param data object whose value is applied to the corresponding function
+ * that is defined in `evolver` at the same path.
+ * @param evolver object that include functions that is applied to
+ * the corresponding value of `data` object at the same path.
+ * @signature
+ *    R.evolve(transf)(data)
+ * @example
+ *    const transf = {
+ *      count: add(1),
+ *      time: { elapsed: add(1), remaining: add(-1) },
+ *    };
+ *    const data = {
+ *      id: 10,
+ *      count: 10,
+ *      time: { elapsed: 100, remaining: 1400 },
+ *    };
+ *    R.pipe(object, R.evolve(transf))
+ *    // => {
+ *    //   id: 10,
+ *    //   count: 11,
+ *    //   time: { elapsed: 101, remaining: 1399 },
+ *    // }
+ * @dataLast
+ * @category Object
+ */
+export function evolve<E extends EvolverStructure>(
+  evolver: E
+): <T extends EvolveTarget<E>>(object: T) => Evolved<T, E>;
+
+export function evolve() {
+  return purry(_evolve, arguments);
+}
+
+function _evolve(data: NestedObject, evolver: EvolverStructure): NestedObject {
```

Comment:
As mentioned above, NestedObject is overkill for this function, simply use `unknown` in the signature and rely on the test below (`typeof data !== "object" || data === null`) to get an `object` which you can spread.

```suggestion
function _evolve(data: unknown, evolver: GenericEvolver): unknown{
```

Reply:
Thank you very much.